File: 003.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if s == '':
             return 0
        mem = {}
        max = 0
        start = 0
        end = 0
        for i, c in enumerate(s):
            if c in mem and mem[c] >= start:
                if end - start + 1 > max:
                    max = end - start + 1
                start = mem[c] + 1
            end = i
            mem[c] = i
        if end - start + 1 > max:
                    max = end - start + 1
        return max

if __name__ == '__main__':
    print(Solution().lengthOfLongestSubstring(' '))

File: 003.cpp
#include <string>
#include <unordered_map>
#include <iostream>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> record;
        int result = 0;
        int start = 0;
        for (int i = 0; i < s.length(); i++) {
            if (record.find(s[i]) != record.end() && record[s[i]] >= start) {
                start = record[s[i]] + 1;
            }
            result = max(result, i-start+1); // [start,i]
            record[s[i]] = i;
        }
        return result;
    }
};

int main() {
    string s = "pwwkew";
    cout << Solution().lengthOfLongestSubstring(s) << endl;
}